plugins/node/instrumentation-fs/src/instrumentation.ts

```diff
@@ -0,0 +1,321 @@
+            this._wrap(
+              fs,
+              fName,
+              <any>this._patchSyncFunction.bind(this, fName)
+            );
+          }
+          return fs;
+        },
+        (fs: FS) => {
+          if (fs === undefined) return;
+          this._diag.debug('Removing patch for fs');
+          if (hasPromises) {
+            for (const fName of PROMISE_FUNCTIONS) {
+              if (isWrapped(fs.promises[fName])) {
+                this._unwrap(fs.promises, fName);
+              }
+            }
+          }
+          for (const fName of ASYNC_FUNCTIONS) {
+            if (isWrapped(fs[fName])) {
+              this._unwrap(fs, fName);
+            }
+          }
+          for (const fName of SYNC_FUNCTIONS) {
+            if (isWrapped(fs[fName])) {
+              this._unwrap(fs, fName);
+            }
+          }
+        }
+      ),
+    ];
+  }
+
+  protected _patchPromiseFunction<T extends (...args: any[]) => ReturnType<T>>(
+    fnName: string,
+    original: T
+  ): T {
+    const instrumentation = this;
+    return <any>async function (this: any, ...args: any[]) {
+      if (isTracingSuppressed(api.context.active())) {
+        // Performance optimization. Avoid creating additional contexts and spans
+        // if we already know that the tracing is being suppressed.
+        return original.apply(this, args);
+      }
+      const { createHook, endHook } =
+        instrumentation.getConfig() as FsInstrumentationConfig;
+      if (typeof createHook === 'function') {
+        if (
+          // promise and async variants get mixed here for the hooks
+          createHook(fnName, {
+            args: args,
+          }) === false
+        ) {
+          // return original.apply(this, args);
+          return api.context.with(
+            suppressTracing(api.context.active()),
+            original,
+            this,
+            ...args
+          );
+        }
+      }
+
+      const span = instrumentation.tracer.startSpan(`fs ${fnName}`) as api.Span;
+      try {
+        // QUESTION: Should we immediately suppress all internal nested calls?
+        const res = await api.context.with(
+          api.trace.setSpan(api.context.active(), span),
+          original,
+          this,
+          ...args
+        );
+        if (typeof endHook === 'function') {
+          try {
+            endHook(fnName, { args: args, span });
+          } catch (e) {
+            instrumentation._diag.error('caught endHook error', e);
+          }
+        }
+        span.end();
+        return res;
+      } catch (err) {
+        span.recordException(err);
+        span.setStatus({
+          message: err.message,
+          code: api.SpanStatusCode.ERROR,
+        });
+        if (typeof endHook === 'function') {
+          endHook(fnName, { args: args, span });
+        }
+        span.end();
+        throw err;
+      }
+    };
+  }
+
+  protected _patchAsyncFunction<T extends (...args: any[]) => ReturnType<T>>(
```

Comment:
I would suggest naming it `_patchCallbackApiFunction`. When I read `Async` I think promises. This is also the name used in nodejs docs

Reply:
Good idea.. Changed that everywhere.